src/libbson/tests/test-json.c

```diff
@@ -3555,6 +3555,86 @@ test_bson_as_json_with_opts_all_types (void)
    bson_destroy (&scope);
 }
 
+static void
+test_overflowing_exponent (void)
+{
+   {
+      bson_decimal128_t decimal128;
+      BSON_ASSERT (!bson_decimal128_from_string ("0E+2147483648", &decimal128));
+   }
+
+   {
+      bson_decimal128_t decimal128;
+      BSON_ASSERT (!bson_decimal128_from_string ("0E-2147483649", &decimal128));
+   }
+
+   {
+      bson_decimal128_t decimal128;
+      BSON_ASSERT (
+         !bson_decimal128_from_string ("-0E+2147483648", &decimal128));
+   }
+   {
+      bson_decimal128_t decimal128;
+      BSON_ASSERT (
+         !bson_decimal128_from_string ("-0E-2147483649", &decimal128));
+   }
+   {
+      bson_error_t error;
+      const char *degenerate_extjson =
+         "{\"d\" : {\"$numberDecimal\" : \"0E+2147483648\"}}";
+
+      BSON_ASSERT (!bson_new_from_json (
+         (const uint8_t *) degenerate_extjson, -1, &error));
+      ASSERT_ERROR_CONTAINS (
+         error,
+         BSON_ERROR_JSON,
+         BSON_JSON_ERROR_READ_INVALID_PARAM,
+         "Invalid input string \"0E+2147483648\", looking for DECIMAL128");
+   }
+
+   {
+      bson_error_t error;
+      const char *degenerate_extjson =
+         "{\"d\" : {\"$numberDecimal\" : \"0E-2147483649\"}}";
+
+      BSON_ASSERT (!bson_new_from_json (
+         (const uint8_t *) degenerate_extjson, -1, &error));
+      ASSERT_ERROR_CONTAINS (
+         error,
+         BSON_ERROR_JSON,
+         BSON_JSON_ERROR_READ_INVALID_PARAM,
+         "Invalid input string \"0E-2147483649\", looking for DECIMAL128");
+   }
+
+   {
+      bson_error_t error;
+      const char *degenerate_extjson =
+         "{\"d\" : {\"$numberDecimal\" : \"-0E+2147483648\"}}";
+
+      BSON_ASSERT (!bson_new_from_json (
+         (const uint8_t *) degenerate_extjson, -1, &error));
+      ASSERT_ERROR_CONTAINS (
+         error,
+         BSON_ERROR_JSON,
+         BSON_JSON_ERROR_READ_INVALID_PARAM,
+         "Invalid input string \"-0E+2147483648\", looking for DECIMAL128");
+   }
+
+   {
+      bson_error_t error;
+      const char *degenerate_extjson =
+         "{\"d\" : {\"$numberDecimal\" : \"-0E-2147483649\"}}";
+
+      BSON_ASSERT (!bson_new_from_json (
+         (const uint8_t *) degenerate_extjson, -1, &error));
+      ASSERT_ERROR_CONTAINS (
+         error,
+         BSON_ERROR_JSON,
+         BSON_JSON_ERROR_READ_INVALID_PARAM,
+         "Invalid input string \"-0E-2147483649\", looking for DECIMAL128");
+   }
```

Comment:
Suggest testing what happens when parsing an exponent outside of range of INT64 min and max.